host/settings/update.go

```diff
@@ -25,67 +33,141 @@ func (cm *ConfigManager) UpdateChainState(tx UpdateStateTx, reverted []chain.Rev
+				return fmt.Errorf("failed to revert last v2 announcement: %w", err)
+			}
+		}
 	}
 
-	var nextAnnouncement *Announcement
+	var announcement Announcement
+	var v2AnnounceAddresses []chain.NetAddress
+	var v2AnnounceIndex types.ChainIndex
 	for _, cau := range applied {
 		index := cau.State.Index
 
-		chain.ForEachHostAnnouncement(cau.Block, func(announcement chain.HostAnnouncement) {
-			if announcement.PublicKey != pk {
+		chain.ForEachHostAnnouncement(cau.Block, func(a chain.HostAnnouncement) {
+			if a.PublicKey != pk {
 				return
 			}
 
-			nextAnnouncement = &Announcement{
-				Address: announcement.NetAddress,
+			announcement = Announcement{
+				Address: a.NetAddress,
 				Index:   index,
 			}
 		})
+
+		chain.ForEachV2HostAnnouncement(cau.Block, func(hostKey types.PublicKey, addresses []chain.NetAddress) {
+			if hostKey != pk {
+				return
+			}
+
+			v2AnnounceAddresses = addresses
+			v2AnnounceIndex = index
+		})
 	}
 
-	if nextAnnouncement == nil {
-		return nil
+	if announcement.Index != (types.ChainIndex{}) {
+		if err := tx.SetLastAnnouncement(announcement); err != nil {
+			return fmt.Errorf("failed to set last announcement: %w", err)
+		}
+		cm.log.Debug("announcement confirmed", zap.String("netaddress", announcement.Address), zap.Stringer("index", announcement.Index))
 	}
 
-	if err := tx.SetLastAnnouncement(*nextAnnouncement); err != nil {
-		return fmt.Errorf("failed to set last announcement: %w", err)
+	if len(v2AnnounceAddresses) > 0 {
+		h := types.NewHasher()
+		types.EncodeSlice(h.E, v2AnnounceAddresses)
+		if err := h.E.Flush(); err != nil {
+			return fmt.Errorf("failed to hash v2 announcement addresses: %w", err)
+		} else if err := tx.SetLastV2AnnouncementHash(h.Sum(), v2AnnounceIndex); err != nil {
+			return fmt.Errorf("failed to set last v2 announcement: %w", err)
+		}
+
+		addresses := make([]string, 0, len(v2AnnounceAddresses))
+		for _, addr := range v2AnnounceAddresses {
+			addresses = append(addresses, fmt.Sprintf("%s/%s", addr.Protocol, addr.Address)) // TODO: Stringer?
+		}
+		cm.log.Debug("v2 announcement confirmed", zap.Strings("addresses", addresses), zap.Stringer("index", v2AnnounceIndex))
 	}
-	cm.log.Debug("announcement confirmed", zap.String("netaddress", nextAnnouncement.Address), zap.Stringer("index", nextAnnouncement.Index))
 	return nil
 }
 
 // ProcessActions processes announcement actions based on the given chain index.
 func (m *ConfigManager) ProcessActions(index types.ChainIndex) error {
-	announcement, err := m.store.LastAnnouncement()
-	if err != nil {
-		return fmt.Errorf("failed to get last announcement: %w", err)
-	}
+	n := m.chain.TipState().Network
+	hostPub := m.hostKey.PublicKey()
 
-	nextHeight := announcement.Index.Height + m.announceInterval
-	netaddress := m.Settings().NetAddress
-	if err := validateNetAddress(netaddress); err != nil {
-		if m.validateNetAddress {
-			return nil
+	// check if there is an unconfirmed announcement
+	var unconfirmedAnnouncement bool
```

Comment:
`nit` no need for the `bool`, might as well return immediately, especially because you only `break` out of the inner loop